File: Packages/MeterKit/Sources/MeterFeatures/Settings/ReminderNotificationCopy.swift
import Foundation

/// 通知文案。
///
/// 边界（不可协商，SPEC 12.5）：
/// 1. 只用本地通知，不接 APNs、不做后台取数。
/// 2. 不做任何基于金额的告警。
/// 3. 通知内容里不出现金额、百分比、provider 名单。
///    触发时 App 没运行，能拿到的只有旧数字；文案只说「去看看」，
///    可以带事实性补充（「上次刷新 3 天前」）。
enum ReminderNotificationCopy {
    static var title: String { String(localized: L("该看一眼了")) }
    static var lookBody: String { String(localized: L("打开 App，看看这个月花了多少。")) }

    static func content(
        lastRefreshAt: Date?,
        now: Date,
        calendar: Calendar
    ) -> ReminderNotificationContent {
        ReminderNotificationContent(
            title: title,
            body: body(lastRefreshAt: lastRefreshAt, now: now, calendar: calendar)
        )
    }

    static func body(
        lastRefreshAt: Date?,
        now: Date,
        calendar: Calendar
    ) -> String {
        guard let lastRefreshAt else { return lookBody }
        let days = calendar.dateComponents(
            [.day],
            from: calendar.startOfDay(for: lastRefreshAt),
            to: calendar.startOfDay(for: now)
        ).day ?? 0
        if days <= 0 {
            return lookBody + String(localized: L("上次刷新就在今天。"))
        }
        if days == 1 {
            return lookBody + String(localized: L("上次刷新 1 天前。"))
        }
        return lookBody + String(localized: L("上次刷新 \(days) 天前。"))
    }
}

File: Android/app/src/main/kotlin/com/zhechengqi/tollcat/settings/ReminderNotificationCopy.kt
package com.zhechengqi.tollcat.settings

import android.content.Context
import com.zhechengqi.tollcat.R
import java.time.Instant
import java.time.ZoneId
import java.time.temporal.ChronoUnit

/**
 * 通知文案。和 iOS `ReminderNotificationCopy` 同一组句子。
 *
 * 不出现金额、百分比、provider 名单（SPEC 12.5）。触发时 App 没运行，
 * 能拿到的只有上次刷新时刻；只说「去看看」，可以带「上次刷新 N 天前」。
 */
object ReminderNotificationCopy {
    fun title(context: Context): String {
        return context.getString(R.string.settings_reminder_notice_title)
    }

    fun lookBody(context: Context): String {
        return context.getString(R.string.settings_reminder_notice_body)
    }

    fun body(
        context: Context,
        lastRefreshAtMillis: Long?,
        nowMillis: Long,
        zoneId: ZoneId = ZoneId.systemDefault(),
    ): String {
        val look = lookBody(context)
        if (lastRefreshAtMillis == null) return look
        val days = daysBetween(lastRefreshAtMillis, nowMillis, zoneId)
        val extra = when {
            days <= 0 -> context.getString(R.string.settings_reminder_notice_today)
            days == 1 -> context.getString(R.string.settings_reminder_notice_yesterday)
            else -> context.getString(R.string.settings_reminder_notice_days, days)
        }
        return look + extra
    }

    internal fun daysBetween(
        fromMillis: Long,
        toMillis: Long,
        zoneId: ZoneId = ZoneId.systemDefault(),
    ): Int {
        val from = Instant.ofEpochMilli(fromMillis).atZone(zoneId).toLocalDate()
        val to = Instant.ofEpochMilli(toMillis).atZone(zoneId).toLocalDate()
        return ChronoUnit.DAYS.between(from, to).toInt()
    }
}
